src/com/rcv/Tabulator.java

```diff
@@ -715,13 +715,32 @@ private boolean hasContinuingCandidates(SortedMap<Integer, Set<String>> rankToCa
       // this is used to determine how many skipped rankings occurred in the case of
       // undervotes
       int lastRank = 0;
-      // loop over all rankings in this CVR from most preferred to least and see how they will
-      // be rendered
+      // candidatesSeen is the set of candidates we've encountered while processing this CVR
+      // in this round; only relevant if exhaustOnDuplicateCandidate is enabled
+      Set<String> candidatesSeen = new HashSet<>();
+      // loop over all rankings in this CVR from most preferred to least until we reach a decision
       // rank iterates through all ranks in this CVR ranking set
       for (int rank : cvr.rankToCandidateIDs.keySet()) {
         // candidateIDSet is all candidates selected at the current rank
         Set<String> candidateIDSet = cvr.rankToCandidateIDs.get(rank);
 
+        // possibly check for a duplicate candidate
+        if (config.isExhaustOnDuplicateCandidateEnabled()) {
+          // the identity of the duplicate candidate, if found
+          String duplicateCandidate = null;
+          for (String candidate : candidateIDSet) {
+            if (candidatesSeen.contains(candidate)) {
+              duplicateCandidate = candidate;
+              break; // finding one duplicate is enough
+            }
+            candidatesSeen.add(candidate);
+          }
+          if (duplicateCandidate != null) {
+            cvr.exhaust("duplicate candidate: " + duplicateCandidate);
+            break;
```

Comment:
Is the intention here that the entire ballot is thrown out completely and not counted, or that it's counted until it sees a duplicate; which means the first time it see it the vote for the duplicate candidate it counts at that level?

If it's the former (which I think makes the most sense), I think we either need to break twice out of here, or place this whole block above the `if (cvr.isExhausted())` check (or add another one below it).

If it's the latter, is there a guarantee that the for loop would traverse `cvr.rankToCandidateIDs.keySet()` in the desired order?

Reply:
It's the latter -- it should be treated normally until the tabulator happens to encounter a duplicate. I don't think I've seen any rule variants yet in which the tabulation is supposed to look ahead for errors; as far as I know, you always consider only as much of the ballot as is necessary to reach a decision in each round.

The for loop is guaranteed to traverse `cvr.rankToCandidateIDs.keySet()` in the right order because `rankToCandidateIDs` is a SortedMap.

Reply:
Ok, so the implicit assumption here is that the first time a duplicate value is encountered is the correct one; thus the bias is towards a higher valuation of the duplicated candidate. I'd argue that the duplicated candidate should be thrown out of the vote entirely (if not the whole vote), but if this is a standard known rule in IRV elections, I guess it is what it is.

Reply:
Yeah. Perhaps other jurisdictions have other variants, but so far this is all that George has mentioned.